artiq/dashboard/datasets.py

```diff
@@ -82,6 +83,69 @@ def get_edit_widget_value(self):
         return self.edit_widget.text()
 
 
+class Creator(QtWidgets.QDialog):
+    def __init__(self, parent, dataset_ctl):
+        QtWidgets.QDialog.__init__(self, parent=parent)
+        self.dataset_ctl = dataset_ctl
+
+        self.setWindowTitle("Create dataset")
+        grid = QtWidgets.QGridLayout()
+        grid.setRowMinimumHeight(1, 35)
+        grid.setColumnMinimumWidth(2, 60)
+        self.setLayout(grid)
+
+        grid.addWidget(QtWidgets.QLabel("Name:"), 0, 0)
+        self.name_widget = QtWidgets.QLineEdit()
+        grid.addWidget(self.name_widget, 0, 1)
+
+        grid.addWidget(QtWidgets.QLabel("Value:"), 1, 0)
+        self.value_widget = QtWidgets.QLineEdit()
+        self.value_widget.setPlaceholderText('PYON (Python)')
+        grid.addWidget(self.value_widget, 1, 1)
+        self.data_type = QtWidgets.QLabel("data type")
+        grid.addWidget(self.data_type, 1, 2)
+        self.value_widget.textChanged.connect(self.dtype)
+
+        grid.addWidget(QtWidgets.QLabel("Persist:"), 2, 0)
+        self.box_widget = QtWidgets.QCheckBox()
+        grid.addWidget(self.box_widget, 2, 1)
+
+        self.ok = QtWidgets.QPushButton('&Ok')
+        self.ok.setEnabled(False)
+        self.cancel = QtWidgets.QPushButton('&Cancel')
+        self.buttons = QtWidgets.QDialogButtonBox(self)
+        self.buttons.addButton(
+            self.ok, QtWidgets.QDialogButtonBox.AcceptRole)
+        self.buttons.addButton(
+            self.cancel, QtWidgets.QDialogButtonBox.RejectRole)
+        grid.setRowStretch(3, 1)
+        grid.addWidget(self.buttons, 4, 0, 1, 3)
+        self.buttons.accepted.connect(self.accept)
+        self.buttons.rejected.connect(self.reject)
+
+    def accept(self):
+        key = self.name_widget.text()
+        value = self.value_widget.text()
+        persist = self.box_widget.isChecked()
+        asyncio.ensure_future(exc_to_warning(self.dataset_ctl.set(
+            key, pyon.decode(value), persist)))
+        QtWidgets.QDialog.accept(self)
+
+    def dtype(self):
+        txt = self.value_widget.text()
+        try:
+            self.data_type.setText(type(pyon.decode(txt)).__name__)
+        except:
+            self.value_widget.setStyleSheet("border: 1px solid red;"
+                                            "height: 28px;"
```

Comment:
Why the explicit `height`? Is this compatible across OSes/UI toolkits?

Reply:
I tested on KDE and XFce, style sheet of `border: 1px solid red` will change the height differently from the default. I fixed the `height` to be the same in KDE but it is indeed different from XFce. So I'll look for another solution.

Reply:
I also tested on GNOME. Changing the border properties of the style sheet will also affect padding and some other properties that make it inconsistent and unaesthetic across OSes. Therefore, I propose to change the background color of the line edit box instead as shown below
![image](https://user-images.githubusercontent.com/43180603/125231649-ba067f00-e30d-11eb-87e0-450d904783eb.png)

Reply:
I'm not sure if a hardcoded color is a good idea, since the user-defined theme colors may change. One obvious problem: You are using the dark theme, but most people use the default with black text and it will be hard to read on a red background.
What about displaying a warning icon at the right of the text box (possibly inside it) when the input is incorrect? Or instead of the N/A text?

Reply:
I changed the `N/A` text into `MessageBoxWarning`. The icon and its size will change according to the OS.
![image](https://user-images.githubusercontent.com/43180603/125250862-4b371f00-e329-11eb-9a2c-6d182f2fdfe0.png)